gui/src/app/CompileContext/CompileContextProvider.tsx

```diff
@@ -0,0 +1,102 @@
+import { ProjectContext } from "@SpCore/ProjectContextProvider";
+import compileStanProgram from "@SpCompileContext/compileStanProgram";
+import {
+  FunctionComponent,
+  PropsWithChildren,
+  useCallback,
+  useContext,
+  useEffect,
+  useState,
+} from "react";
+import { CompileContext, CompileStatus } from "./CompileContext";
+
+type CompileContextProviderProps = {
+  // none
+};
+
+const initialStanWasmServerUrl =
+  localStorage.getItem("stanWasmServerUrl") ||
+  "https://trom-stan-wasm-server.magland.org";
+
+export const CompileContextProvider: FunctionComponent<
+  PropsWithChildren<CompileContextProviderProps>
+> = ({ children }) => {
+  const { data } = useContext(ProjectContext);
+  const [compileStatus, setCompileStatus] = useState<CompileStatus>("");
+  const [
+    theStanFileContentThasHasBeenCompiled,
+    setTheStanFileContentThasHasBeenCompiled,
+  ] = useState<string>("");
+  const [compileMessage, setCompileMessage] = useState<string>("");
+  const [compiledMainJsUrl, setCompiledMainJsUrl] = useState<
+    string | undefined
+  >(undefined);
+  const [validSyntax, setValidSyntax] = useState<boolean>(false);
+
+  useEffect(() => {
+    // if the compiled content is not the same as the current content,
+    // then the state should not be compiled or failed
+    if (data.stanFileContent !== theStanFileContentThasHasBeenCompiled) {
+      if (compileStatus === "compiled" || compileStatus === "failed") {
+        setCompileStatus("");
+        setCompiledMainJsUrl("");
+      }
+    }
```

Comment:
Is it worthwhile to do the flip side of this--such that if the `stanFileContent` is flipped back to equal the content that has been compiled, we revert to the "compiled" state?

I'm imagining deleting and re-adding a semicolon for instance.

Reply:
I'd vote against that because (a) it would be challenging to manage that -- you see in this code the compiledMainJsUrl gets cleared (b) it's easy enough to click the compile button again, and it will be a cache hit.